onnxruntime/core/optimizer/nchwc_transformer.cc

```diff
@@ -666,6 +666,125 @@ void NchwcTransformerImpl::TransformActivation(Node& node) {
+    return;
+  }
+  auto* nchwc_input = it->second.get();
+
+  // Require that BatchNormalization-7 uses spatial normalization.
+  const auto* spatial_attr = graph_utils::GetNodeAttribute(node, "spatial");
+  if (spatial_attr != nullptr && utils::HasInt(*spatial_attr) && spatial_attr->i() != 1) {
+    return;
+  }
+
+  const auto* epsilon_attr = graph_utils::GetNodeAttribute(node, "epsilon");
+  if (epsilon_attr == nullptr || !utils::HasFloat(*epsilon_attr)) {
+    return;
+  }
+  float epsilon = static_cast<float>(epsilon_attr->f());
+
+  const int64_t channels = nchwc_input->channels_;
+
+  auto get_bn_tensor_proto = [this, channels](const std::string& input_name) {
+    const auto* tensor_proto = graph_utils::GetConstantInitializer(graph_, input_name);
+    if (tensor_proto != nullptr) {
+      if ((tensor_proto->data_type() != ONNX_NAMESPACE::TensorProto_DataType_FLOAT) ||
+          (tensor_proto->dims_size() != 1) ||
+          (tensor_proto->dims(0) != channels)) {
+        tensor_proto = nullptr;
+      }
+    }
+    return tensor_proto;
+  };
+
+  const auto* bn_scale_tensor_proto = get_bn_tensor_proto(input_defs[1]->Name());
+  if (bn_scale_tensor_proto == nullptr) {
+    return;
+  }
+  const auto* bn_B_tensor_proto = get_bn_tensor_proto(input_defs[2]->Name());
+  if (bn_B_tensor_proto == nullptr) {
+    return;
+  }
+  const auto* bn_mean_tensor_proto = get_bn_tensor_proto(input_defs[3]->Name());
+  if (bn_mean_tensor_proto == nullptr) {
+    return;
+  }
+  const auto* bn_var_tensor_proto = get_bn_tensor_proto(input_defs[4]->Name());
+  if (bn_var_tensor_proto == nullptr) {
+    return;
+  }
+
+  auto bn_scale = onnxruntime::make_unique<Initializer>(*bn_scale_tensor_proto);
+  auto bn_B = onnxruntime::make_unique<Initializer>(*bn_B_tensor_proto);
+  auto bn_mean = onnxruntime::make_unique<Initializer>(*bn_mean_tensor_proto);
+  auto bn_var = onnxruntime::make_unique<Initializer>(*bn_var_tensor_proto);
+
+  // Calculate the scale and bias for the replacement convolution.
+  bn_var->add(epsilon);
+  bn_var->sqrt();
+  bn_scale->div(*bn_var);
+  bn_mean->mul(*bn_scale);
+  bn_B->sub(*bn_mean);
+
+  const size_t nchwc_block_size = MlasNchwcGetBlockSize();
+  const int64_t nchwc_channels = (channels + nchwc_block_size - 1) & ~(nchwc_block_size - 1);
+
+  std::vector<float> padded_buffer(nchwc_channels);
+
+  std::copy_n(bn_scale->data<float>(), channels, padded_buffer.data());
+
+  ONNX_NAMESPACE::TensorProto nchwc_conv_W_tensor_proto;
+  nchwc_conv_W_tensor_proto.set_data_type(ONNX_NAMESPACE::TensorProto_DataType_FLOAT);
+  nchwc_conv_W_tensor_proto.set_name(graph_.GenerateNodeArgName("bn_scale"));
+  nchwc_conv_W_tensor_proto.set_raw_data(padded_buffer.data(), nchwc_channels * sizeof(float));
+  nchwc_conv_W_tensor_proto.add_dims(nchwc_channels);
+  nchwc_conv_W_tensor_proto.add_dims(1);
+  nchwc_conv_W_tensor_proto.add_dims(1);
+  nchwc_conv_W_tensor_proto.add_dims(1);
+
+  graph_.AddInitializedTensor(nchwc_conv_W_tensor_proto);
+
```

Comment:
graph_utils::AddInitializer would be slightly preferable.

Reply:
Thanks, I updated the transformer to use this helper instead.